client/driver/rkt.go

```diff
@@ -563,13 +727,56 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
 				return nil, fmt.Errorf("port_map is not set. When you defined port in the resources, you need to configure port_map.")
 			}
 
-			hostPortStr := strconv.Itoa(port.Value)
-
 			d.logger.Printf("[DEBUG] driver.rkt: exposed port %s", containerPort)
+
 			// Add port option to rkt run arguments. rkt allows multiple port args
-			prepareArgs = append(prepareArgs, fmt.Sprintf("--port=%s:%s", containerPort, hostPortStr))
+			if img != "" {
+				hostPortStr := strconv.Itoa(port.Value)
+				prepareArgs = append(prepareArgs, fmt.Sprintf("--port=%s:%s", containerPort, hostPortStr))
+			} else {
+				containerPortACName, err := appctypes.NewACName(containerPort)
+				if err != nil {
+					return nil, fmt.Errorf("[ERR] driver.rkt: failed to set ACName: %s", err)
+				}
+
+				podManifest.Ports = append(podManifest.Ports, appctypes.ExposedPort{
+					Name:     *containerPortACName,
+					HostPort: uint(port.Value),
+				})
+			}
+		}
+	}
+
+	if img == "" {
+		podManifestStr, err := json.Marshal(podManifest)
+		if err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: failed to marshal rkt pod manifest to JSON: %s", err)
+		}
+
+		// Ensure that the generated pod manifest is valid according to the appc spec
+		validatePodManifest := appcschema.PodManifest{}
+		if err := json.Unmarshal(podManifestStr, &validatePodManifest); err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: Invalid rkt pod manifest: %s", err)
+		}
+
+		// Create temporary file to dump the contents of podManifest
+		tmpfile, err := ioutil.TempFile("", "pod")
+		if err != nil {
+			return nil, err
+		}
+
+		podManifestTmpfile := tmpfile.Name()
+		defer os.Remove(podManifestTmpfile)
+		defer tmpfile.Close()
```

Comment:
I'd avoid deferring the `Close()` call and do it as a regular function call after writing to the file so we make sure the file is closed before rkt uses it and we catch possible errors when closing.